File: sylee723/7주차/PGS_도넛과_막대_그래프.java
import java.util.*;

class Solution {
    static Map<Integer, ArrayList<Integer>> outMap;
    
    public int[] solution(int[][] edges) {
        outMap = new HashMap<>();
        Set<Integer> notCreatedSet = new HashSet<>();
        
        for (int i = 0; i < edges.length; i++) {
            if (outMap.containsKey(edges[i][0])) {
                ArrayList<Integer> eList = outMap.get(edges[i][0]);
                eList.add(edges[i][1]);
                outMap.put(edges[i][0], eList);
            } else {
                ArrayList<Integer> eList = new ArrayList<>();
                eList.add(edges[i][1]);
                outMap.put(edges[i][0], eList);
            }
            
            notCreatedSet.add(edges[i][1]);
        }
        
        int created = 0;
        for (int key : outMap.keySet()) {
            if (outMap.get(key).size() >= 2 && !notCreatedSet.contains(key)) {
                created = key;
                break;
            }
        }
        
        int[] answer = new int[4];
        answer[0] = created;

        for (int v : outMap.get(created)) {
            int idx = graphType(v);
            answer[idx]++;
        }
        
        return answer;
    }
    
    static int graphType(int start) {
        // 도넛 1, 막대 2, 8자 3
        int now = start;
        while (true) {
            if (outMap.get(now) != null && outMap.get(now).size() == 2) {
                return 3;
            }
            
            if (outMap.get(now) == null) {
                return 2;
            }
            
            now = outMap.get(now).get(0);
            if (now == start) {
                return 1;
            }
        }
    }
}

File: daehwa_park/7주차/PGS_도넛과_막대그래프.py
def solution(edges):
    answer = [0, 0, 0, 0]
    
    nodeCnt = {}
    
    for s, e in edges:
        if not nodeCnt.get(s):
            nodeCnt[s] = [0, 0]
        if not nodeCnt.get(e):
            nodeCnt[e] = [0, 0]

        nodeCnt[s][0] += 1
        nodeCnt[e][1] += 1
    
    for key, node in nodeCnt.items():

        if node[0] >= 2 and node[1] == 0:
            answer[0] = key
        elif node[0] == 0 and node[1] > 0:
            answer[2] += 1
        elif node[0] >= 2 and node[1] >= 2:
            answer[3] += 1
            
    answer[1] = (nodeCnt[answer[0]][0] - answer[2] - answer[3])

    return answer